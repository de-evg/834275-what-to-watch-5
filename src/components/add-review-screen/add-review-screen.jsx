import React, {PureComponent} from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import {typesMap} from "../../prop-types/prop-types";

const reviewLength = {
  MIN: 50,
  MAX: 400
};

class AddReviewScreen extends PureComponent {
  constructor(props) {
    super(props);
  }

  handleFormSubmit(evt) {
    evt.preventDefault();
  }

  render() {
    const {movies} = this.props;
    const currentMovie = movies.find((movie) => movie.id === +id);
    const {title, previewURL, posterURL} = currentMovie;
    const {renderRatingStars, renderReviewText, textReview, rating, match: {params: {id}}} = this.props;

    return (
      <section className="movie-card movie-card--full">
        <div className="movie-card__header">
          <div className="movie-card__bg">
            <img src={previewURL} alt={title} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <header className="page-header">
            <div className="logo">
              <Link to="/" className="logo__link">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </Link>
            </div>

            <nav className="breadcrumbs">
              <ul className="breadcrumbs__list">
                <li className="breadcrumbs__item">
                  <Link
                    to={`/films/${id}`}
                    className="breadcrumbs__link">{title}
                  </Link>
                </li>
                <li className="breadcrumbs__item">
                  <a className="breadcrumbs__link">Add review</a>
                </li>
              </ul>
            </nav>

            <div className="user-block">
              <div className="user-block__avatar">
                <img src="/img/avatar.jpg" alt="User avatar" width="63" height="63" />
              </div>
            </div>
          </header>

          <div className="movie-card__poster movie-card__poster--small">
            <img src={posterURL} alt={title} width="218" height="327" />
          </div>
        </div>

        <div className="add-review">
          <form action="#" className="add-review__form" onSubmit={this.handleFormSubmit}>
            <div className="rating">
              {renderRatingStars(rating)}
            </div>

            <div className="add-review__text">
              {renderReviewText(textReview)}
              <div className="add-review__submit">
                {
                  <button
                    className="add-review__btn"
                    type="submit"
                    disabled={textReview.length < reviewLength.MIN || textReview.length >= reviewLength.MAX}
                  >
                    Post
                  </button>
                }
              </div>

            </div>
          </form>
        </div>

      </section>
    );
  }
}

AddReviewScreen.propTypes = {
  match: typesMap.match,
  renderRatingStars: typesMap.renderRatingStars,
  renderReviewText: typesMap.renderReviewText,
  textReview: typesMap.textReview,
  rating: typesMap.rating,
  movies: typesMap.movies
};

const mapStateToProps = (state) => ({
  movies: state.DATA.movies
});

export {AddReviewScreen};
export default connect(mapStateToProps)(AddReviewScreen);
