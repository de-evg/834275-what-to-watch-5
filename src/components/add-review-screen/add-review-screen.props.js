import PropTypes from "prop-types";

const AddReviewScreenProps = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string
    }),
  }),
  history: PropTypes.object.isRequired,
  renderRatingStars: PropTypes.func.isRequired,
  renderReviewText: PropTypes.func.isRequired,
  textReview: PropTypes.string.isRequired,
  rating: PropTypes.string.isRequired,
  movie: PropTypes.oneOfType([PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    release: PropTypes.number.isRequired,
    posterURL: PropTypes.string.isRequired,
    previewURL: PropTypes.string.isRequired,
    rating: PropTypes.string.isRequired,
    ratingLevel: PropTypes.string.isRequired,
    ratingCount: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    actors: PropTypes.array.isRequired,
    videoURL: PropTypes.string.isRequired,
    isInWatchList: PropTypes.bool.isRequired
  }), PropTypes.object]),
  onFormSubmit: PropTypes.func.isRequired,
  onReviewReset: PropTypes.func.isRequired,
  user: PropTypes.shape({
    authorizationStatus: PropTypes.string.isRequired,
    userID: PropTypes.number.isRequired,
    userName: PropTypes.string.isRequired,
    userAvatar: PropTypes.string.isRequired,
    userEmail: PropTypes.string.isRequired,
    authError: PropTypes.bool.isRequired
  }),
  loadMovie: PropTypes.func.isRequired,
  movieIsLoaded: PropTypes.bool.isRequired
};

export default AddReviewScreenProps;
