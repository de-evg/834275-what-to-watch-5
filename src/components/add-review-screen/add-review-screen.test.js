import React from "react";
import {BrowserRouter} from "react-router-dom";
import renderer from "react-test-renderer";
import {AddReviewScreen} from "./add-review-screen";

const TextReview = {
  MIN_LENGTH: ``,
  MAX_LENGTH: new Array(41).fill(`0123456789`).join(``),
  MIDDLE_LENGTH: new Array(20).fill(`0123456789`).join(``),
};

describe(`Render AddReviewScreen`, () => {
  it(`Should AddreviewScreen with disabled (min length) submit button render correctly`, () => {
    const tree = renderer
      .create(<BrowserRouter>
        <AddReviewScreen
          match={
            {
              params: {id: ``}
            }
          }
          history={{
            push: () => {}
          }}
          renderRatingStars={() => {}}
          renderReviewText={() => {}}
          textReview={TextReview.MIN_LENGTH}
          rating={``}
          movie={
            {
              title: ``,
              previewURL: ``,
              posterURL: ``,
              id: 0
            }
          }
          user={
            {
              authorizationStatus: `AUTH`,
              userAvatar: ``,
              userID: 0,
              userName: ``,
              userEmail: ``,
              authError: false
            }
          }
          loadMovie={() => {}}
          onReviewReset={() => {}}
          onFormSubmit={() => {}}
        />
      </BrowserRouter>, {
        createNodeMock: () => {
          return {};
        }
      })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should AddreviewScreen with disabled (max length) submit button render correctly`, () => {
    const tree = renderer
      .create(<BrowserRouter>
        <AddReviewScreen
          match={
            {
              params: {id: ``}
            }
          }
          history={{
            push: () => {}
          }}
          renderRatingStars={() => {}}
          renderReviewText={() => {}}
          textReview={TextReview.MAX_LENGTH}
          rating={``}
          movie={
            {
              title: ``,
              previewURL: ``,
              posterURL: ``,
              id: 0
            }
          }
          user={
            {
              authorizationStatus: `AUTH`,
              userAvatar: ``,
              userID: 0,
              userName: ``,
              userEmail: ``,
              authError: false
            }
          }
          loadMovie={() => {}}
          onReviewReset={() => {}}
          onFormSubmit={() => {}}
        />
      </BrowserRouter>, {
        createNodeMock: () => {
          return {};
        }
      })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should AddreviewScreen with enabled submit button render correctly`, () => {
    const tree = renderer
      .create(<BrowserRouter>
        <AddReviewScreen
          match={
            {
              params: {id: ``}
            }
          }
          history={{
            push: () => {}
          }}
          renderRatingStars={() => {}}
          renderReviewText={() => {}}
          textReview={TextReview.MIDDLE_LENGTH}
          rating={`5`}
          movie={
            {
              title: ``,
              previewURL: ``,
              posterURL: ``,
              id: 0
            }
          }
          user={
            {
              authorizationStatus: `AUTH`,
              userAvatar: ``,
              userID: 0,
              userName: ``,
              userEmail: ``,
              authError: false
            }
          }
          loadMovie={() => {}}
          onReviewReset={() => {}}
          onFormSubmit={() => {}}
        />
      </BrowserRouter>, {
        createNodeMock: () => {
          return {};
        }
      })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
