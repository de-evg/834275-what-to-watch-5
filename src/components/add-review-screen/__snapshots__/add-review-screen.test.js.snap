// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render AddReviewScreen Should AddreviewScreen with disabled (max length) submit button render correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt=""
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/0"
              onClick={[Function]}
            >
              
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <a
            href="/myList"
            onClick={[Function]}
          >
            <img
              alt="User avatar"
              height="63"
              src=""
              width="63"
            />
          </a>
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt=""
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      />
      <div
        className="add-review__text"
      >
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={true}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;

exports[`Render AddReviewScreen Should AddreviewScreen with disabled (min length) submit button render correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt=""
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/0"
              onClick={[Function]}
            >
              
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <a
            href="/myList"
            onClick={[Function]}
          >
            <img
              alt="User avatar"
              height="63"
              src=""
              width="63"
            />
          </a>
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt=""
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      />
      <div
        className="add-review__text"
      >
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={true}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;

exports[`Render AddReviewScreen Should AddreviewScreen with enabled submit button render correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt=""
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/0"
              onClick={[Function]}
            >
              
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <a
            href="/myList"
            onClick={[Function]}
          >
            <img
              alt="User avatar"
              height="63"
              src=""
              width="63"
            />
          </a>
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt=""
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      />
      <div
        className="add-review__text"
      >
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={false}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
